From 4db3b4029278b336eb53b96789e03d86bf32fa8e Mon Sep 17 00:00:00 2001
From: Zeng Heng <zengheng4@huawei.com>
Date: Mon, 27 Oct 2025 09:32:24 +0800
Subject: [PATCH 783/960] downstream: arm64/mpam: Fix L2 MBWU monitor
 multiplexing issue

To support the expanded quantities of PARTIDs and PMGs, the MBWU monitors
need to multiplex. Therefore, when a monitor tracks a new PARTID or PMG,
which means a change in monitor configuration is detected, the accumulated
correction should set to zero.

Fixes: 21ac9fc8f275 ("arm_mpam: Track bandwidth counter state for overflow and power management")
Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Quanmin Yan <yanquanmin1@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/platform/mpam/mpam_devices.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/platform/mpam/mpam_devices.c b/drivers/platform/mpam/mpam_devices.c
index 3ade552d1d28..00de37186fa3 100644
--- a/drivers/platform/mpam/mpam_devices.c
+++ b/drivers/platform/mpam/mpam_devices.c
@@ -1013,7 +1013,7 @@ static void __ris_msmon_read(void *arg)
 	bool reset_on_next_read = false;
 	struct mpam_msc_ris *ris = m->ris;
 	struct mpam_msc *msc = m->ris->msc;
-	struct msmon_mbwu_state *mbwu_state;
+	struct msmon_mbwu_state *mbwu_state = NULL;
 	u32 mon_sel, ctl_val, flt_val, cur_ctl, cur_flt;
 
 	lockdep_assert_held(&msc->lock);
@@ -1045,8 +1045,14 @@ static void __ris_msmon_read(void *arg)
 	config_mismatch = cur_flt != flt_val ||
 			  cur_ctl != (ctl_val | MSMON_CFG_x_CTL_EN);
 
-	if (config_mismatch || reset_on_next_read)
+	if (config_mismatch || reset_on_next_read) {
 		write_msmon_ctl_flt_vals(m, ctl_val, flt_val);
+		if (mbwu_state) {
+			mbwu_state->prev_val = 0;
+			mbwu_state->correction = 0;
+			mbwu_overflow = false;
+		}
+	}
 
 	/*
 	 * Selects the monitor instance associated to the specified PARTID
-- 
2.43.0

